base/promotion.jl

```diff
@@ -307,19 +314,64 @@ promote_type(T) = T
 promote_type(T, S, U) = (@inline; promote_type(promote_type(T, S), U))
 promote_type(T, S, U, V...) = (@inline; afoldl(promote_type, promote_type(T, S, U), V...))
 
-promote_type(::Type{Bottom}, ::Type{Bottom}) = Bottom
-promote_type(::Type{T}, ::Type{T}) where {T} = T
-promote_type(::Type{T}, ::Type{Bottom}) where {T} = T
-promote_type(::Type{Bottom}, ::Type{T}) where {T} = T
+const _promote_type_binary_recursion_depth_limit_exception = let
+    s = "`promote_type`: recursion depth limit reached, giving up; check for faulty/conflicting/missing `promote_rule` methods"
+    ArgumentError(s)
+end
+const _promote_type_binary_detected_infinite_recursion_exception = let
+    s = "`promote_type`: detected unbounded recursion caused by faulty `promote_rule` logic"
+    ArgumentError(s)
+end
+function _promote_type_binary(::Type{T}, ::Type{S}, recursion_depth_limit::Tuple{Vararg{Nothing}}) where {T,S}
+    function err_giving_up()
+        @noinline
+        throw(_promote_type_binary_recursion_depth_limit_exception)
+    end
+    function err_detected_infinite_recursion()
+        @noinline
+        throw(_promote_type_binary_detected_infinite_recursion_exception)
+    end
+    type_is_bottom(::Type{X}) where {X} = X === Bottom
+    normalize_type(::Type{X}) where {X} = X
+    detect_loop(::Type{A}, ::Type{B}) where {A, B} = _types_are_equal(T, A) && _types_are_equal(S, B)
+    if type_is_bottom(T)
+        return S
+    end
+    if type_is_bottom(S) || _types_are_equal(S, T)
+        return T
+    end
+    # Try promote_rule in both orders.
+    st = normalize_type(promote_rule(S, T))
+    ts = normalize_type(promote_rule(T, S))
+    # If no promote_rule is defined, both directions give Bottom. In that
+    # case use typejoin on the original types instead.
+    if type_is_bottom(st) && type_is_bottom(ts)
+        return normalize_type(typejoin(T, S))
+    end
+    if detect_loop(ts, st) || detect_loop(st, ts)
+        # This is not strictly necessary, as we already limit the recursion depth, but
+        # makes for nicer UX.
+        err_detected_infinite_recursion()
+    end
+    if recursion_depth_limit === ()
+        err_giving_up()
+    end
+    l = tail(recursion_depth_limit)
+    _promote_type_binary(ts, st, l)
+end
+
+"""
+    _promote_type_binary_recursion_depth_limit::Tuple{Vararg{Nothing}}
+
+Recursion depth limit for `_promote_type_binary`, to prevent stack overflow.
+"""
+const _promote_type_binary_recursion_depth_limit = let
+    n2 = (nothing, nothing)
+    (n2..., n2...)
+end
```

Comment:
Currently the recursion depth limit is set to **four** here.

Idea for a new experiment: run PkgEval again with this low depth limit, if that turns out OK it's a strong signal that this minor change will be nonbreaking in practice after increasing the limit to, e.g., **eight**.